src/keepass_fmt_plug.c

```diff
@@ -37,115 +42,240 @@ john_register_one(&fmt_KeePass);
+	// Add the keyfile, hash again to get the composite key
+	if (cur_salt->have_keyfile) {
 		SHA256_Init(&ctx);
 		SHA256_Update(&ctx, hash, 32);
+		SHA256_Update(&ctx, cur_salt->keyfile, 32);
 		SHA256_Final(hash, &ctx);
-	}
-
-	if (cur_salt->have_keyfile) {
+	} else if (cur_salt->kdbx_ver > 1) {
 		SHA256_Init(&ctx);
 		SHA256_Update(&ctx, hash, 32);
-		SHA256_Update(&ctx, cur_salt->keyfile, 32);
 		SHA256_Final(hash, &ctx);
 	}
 
-	// Next, encrypt the created hash
-	AES_set_encrypt_key(csp->transf_randomseed, 256, &akey);
-	i = csp->key_transf_rounds >> 2;
-	while (i--) {
-		AES_encrypt(hash, hash, &akey);
-		AES_encrypt(hash, hash, &akey);
-		AES_encrypt(hash, hash, &akey);
-		AES_encrypt(hash, hash, &akey);
-		AES_encrypt(hash+16, hash+16, &akey);
-		AES_encrypt(hash+16, hash+16, &akey);
-		AES_encrypt(hash+16, hash+16, &akey);
-		AES_encrypt(hash+16, hash+16, &akey);
-	}
-	i = csp->key_transf_rounds & 3;
-	while (i--) {
-		AES_encrypt(hash, hash, &akey);
-		AES_encrypt(hash+16, hash+16, &akey);
-	}
+	// Next, encrypt the composite key to get the transformed key (only for AES-KDF)
+	if (cur_salt->kdf == 0) {
+		AES_KEY akey;
+		AES_set_encrypt_key(cur_salt->transf_randomseed, 256, &akey);
 
-	// Finally, hash it again...
-	SHA256_Init(&ctx);
-	SHA256_Update(&ctx, hash, 32);
-	SHA256_Final(hash, &ctx);
+		unsigned int rounds = cur_salt->key_transf_rounds;
+		while (rounds--) {
+			AES_encrypt(hash, hash, &akey);
+			AES_encrypt(hash+16, hash+16, &akey);
+		}
 
-	// ...and hash the result together with the random seed
-	SHA256_Init(&ctx);
-	if (csp->version == 1) {
-		SHA256_Update(&ctx, csp->final_randomseed, 16);
-	}
-	else {
-		SHA256_Update(&ctx, csp->final_randomseed, 32);
+		// Finally, hash it again to get the master key
+		SHA256_Init(&ctx);
+		SHA256_Update(&ctx, hash, 32);
+		SHA256_Final(final_key, &ctx);
+
+		if (cur_salt->kdbx_ver < 4) {
+			// ...and hash the result together with the random seed
+			SHA256_Init(&ctx);
+			if (cur_salt->kdbx_ver == 1)
+				SHA256_Update(&ctx, cur_salt->final_randomseed, 16);
+			else
+				SHA256_Update(&ctx, cur_salt->final_randomseed, 32);
+			SHA256_Update(&ctx, final_key, 32);
+			SHA256_Final(final_key, &ctx);
+		}
+	} else if (cur_salt->kdf == 1) { // Argon2
+		argon2_error_codes error_code;
+		argon2_context context = {
+			.out = (uint8_t*)final_key,
+			.outlen = 32,
+			.pwd = (uint8_t*)hash,
+			.pwdlen = 32,
+			.salt = (uint8_t*)cur_salt->transf_randomseed,
+			.saltlen = 32,
+			.t_cost = cur_salt->t_cost,
+			.m_cost = cur_salt->m_cost,
+			.lanes = cur_salt->lanes,
+			.threads = cur_salt->lanes,
+			.allocate_cbk = &allocate,
+			.free_cbk = &deallocate,
+			.flags = ARGON2_DEFAULT_FLAGS,
+			.version = ARGON2_VERSION_13
+		};
+		error_code = argon2_ctx(&context, cur_salt->type);
+		if (error_code != ARGON2_OK) {
+			ret = -1;
+			fprintf(stderr, "Keepass: Error: Argon2 failed: %s\n",
+			        argon2_error_message(error_code));
```

Comment:
Other error messages in this source file (in the allocation functions) start with `Error: KeePass: `, so let's make this one similar.